src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/RazorCustomMessageTarget.cs

```diff
@@ -143,4 +163,127 @@ internal partial class RazorCustomMessageTarget
     }
 
     private record struct DelegationRequestDetails(string LanguageServerName, Uri ProjectedUri, ITextBuffer TextBuffer);
+
+    private async Task<SynchronizedResult<TVirtualDocumentSnapshot>> TrySynchronizeVirtualDocumentAsync<TVirtualDocumentSnapshot>(
+       int requiredHostDocumentVersion,
+       TextDocumentIdentifier hostDocument,
+       CancellationToken cancellationToken,
+       bool rejectOnNewerParallelRequest = true)
+       where TVirtualDocumentSnapshot : VirtualDocumentSnapshot
+    {
+        // For Html documents we don't do anything fancy, just call the standard service
+        // If we're not generating unique document file names, then we can treat C# documents the same way
+        if (!_languageServerFeatureOptions.IncludeProjectKeyInGeneratedFilePath ||
+            typeof(TVirtualDocumentSnapshot) == typeof(HtmlVirtualDocumentSnapshot))
+        {
+            return await _documentSynchronizer.TrySynchronizeVirtualDocumentAsync<TVirtualDocumentSnapshot>(requiredHostDocumentVersion, hostDocument.Uri, cancellationToken).ConfigureAwait(false);
+        }
+
+        var virtualDocument = FindVirtualDocument<TVirtualDocumentSnapshot>(hostDocument.Uri, hostDocument.GetProjectContext());
+
+        if (virtualDocument is { ProjectKey.Id: null })
+        {
+            _outputWindowLogger?.LogDebug("Trying to sync to a doc with no project Id. Waiting 500ms for document add.");
+            if (await _csharpVirtualDocumentAddListener.WaitForDocumentAddAsync(TimeSpan.FromMilliseconds(500), cancellationToken).ConfigureAwait(false))
```

Comment:
Is there a reason 500ms was specifically chosen? It looks like the UpdateCSharpBuffer class hardcodes in 500ms as well. Would it be better to make 500ms the default value in the listener method, and then give callers an ability to override it with a different value?

Reply:
Nope, pure guesswork. 100ms was too short for the speed of the integration tests 🤷‍♂️